code/datums/components/riding.dm

```diff
@@ -0,0 +1,309 @@
+	last_vehicle_move = world.time
+
+	if(keycheck(user))
+		var/turf/next = get_step(AM, direction)
+		var/turf/current = get_turf(AM)
+		if(!istype(next) || !istype(current))
+			return	//not happening.
+		if(!turf_check(next, current))
+			to_chat(user, "<span class='warning'>Your \the [AM] can not go onto [next]!</span>")
+			return
+		if(!isturf(AM.loc))
+			return
+		step(AM, direction)
+
+		if((direction & (direction - 1)) && (AM.loc == next))		//moved diagonally
+			last_move_diagonal = TRUE
+		else
+			last_move_diagonal = FALSE
+
+		handle_vehicle_layer()
+		handle_vehicle_offsets()
+	else
+		to_chat(user, "<span class='warning'>You'll need the keys in one of your hands to [drive_verb] [AM].</span>")
+
+/datum/component/riding/proc/Unbuckle(atom/movable/buckled_thing)
+	addtimer(CALLBACK(parent, /atom/movable/.proc/unbuckle_mob, buckled_thing), 0, TIMER_UNIQUE)
+
+///////Yes, I said humans. No, this won't end well...//////////
+/datum/component/riding/human
+	del_on_unbuckle_all = TRUE
+
+/datum/component/riding/human/Initialize()
+	. = ..()
+	RegisterSignal(parent, COMSIG_HUMAN_MELEE_UNARMED_ATTACK, .proc/on_carrier_unarmed_melee)
+
+/datum/component/riding/human/vehicle_mob_unbuckle(datum/source, mob/living/buckled_mob, force = FALSE)
+	var/mob/living/carbon/human/human_carrier = parent
+	human_carrier.remove_movespeed_modifier(MOVESPEED_ID_HUMAN_CARRYING)
+	UnregisterSignal(buckled_mob, COMSIG_MOVABLE_PRE_THROW)
+	return ..()
+
+/datum/component/riding/human/vehicle_mob_buckle(datum/source, mob/living/buckling_mob, force = FALSE)
+	. = ..()
+	var/mob/living/carbon/human/human_carrier = parent
+	human_carrier.add_movespeed_modifier(MOVESPEED_ID_HUMAN_CARRYING, TRUE, 0, NONE, TRUE, HUMAN_CARRY_SLOWDOWN)
+	RegisterSignal(buckling_mob, COMSIG_MOVABLE_PRE_THROW, .proc/on_passenger_throw)
+
+
+/datum/component/riding/human/proc/on_carrier_unarmed_melee(datum/source, atom/target)
+	var/mob/living/carbon/human/human_carrier = parent
+	if(human_carrier.a_intent != INTENT_DISARM || !(target in human_carrier.buckled_mobs))
+		return
+	force_dismount(target)
+
+/datum/component/riding/human/proc/on_passenger_throw(datum/source, atom/target, range, speed, thrower, spin)
+	force_dismount(source, TRUE)
+
+
+/datum/component/riding/human/handle_vehicle_layer()
+	var/atom/movable/AM = parent
+	if(AM.buckled_mobs && AM.buckled_mobs.len)
+		for(var/mob/M in AM.buckled_mobs) //ensure proper layering of piggyback and carry, sometimes weird offsets get applied
+			M.layer = MOB_LAYER
+		if(!AM.buckle_lying)
+			if(AM.dir == SOUTH)
+				AM.layer = ABOVE_MOB_LAYER
+			else
+				AM.layer = OBJ_LAYER
+		else
+			if(AM.dir == NORTH)
+				AM.layer = OBJ_LAYER
+			else
+				AM.layer = ABOVE_MOB_LAYER
+	else
+		AM.layer = MOB_LAYER
+
+/datum/component/riding/human/get_offsets(pass_index)
+	var/mob/living/carbon/human/H = parent
+	if(H.buckle_lying)
+		return list(TEXT_NORTH = list(0, 6), TEXT_SOUTH = list(0, 6), TEXT_EAST = list(0, 6), TEXT_WEST = list(0, 6))
```

Comment:
what are these magic numbers, offsets?

Reply:
Yea, the proc is called `get_offsets()`
Not my code, but not sure how that'd be more clear.